src/block_device.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/*!
+Block device helper functions.
+*/
+
+use snafu::{ensure, ResultExt, Snafu};
+use std::convert::TryFrom;
+use std::fs::OpenOptions;
+use std::os::unix::io::AsRawFd;
+use std::path::Path;
+
+#[derive(Debug, Snafu)]
+pub struct Error(error::Error);
+type Result<T> = std::result::Result<T, Error>;
+
+/// Generates the blkgetsize64 function.
+mod ioctl {
+    use nix::ioctl_read;
+    ioctl_read!(blkgetsize64, 0x12, 114, u64);
+}
+
+/// Find the size of a block device.
+pub(crate) fn get_block_device_size(path: &Path) -> Result<i64> {
+    let file = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .context(error::OpenFile { path })?;
+
+    let mut block_device_size = 0;
+    let result = unsafe { ioctl::blkgetsize64(file.as_raw_fd(), &mut block_device_size) }
+        .context(error::GetBlockDeviceSize)?;
```

Comment:
This error variant doesn't include the path, which seems like it'd always be helpful with an IO error, but the wrapping error in upload.rs does include it - maybe we add the path to this variant, and use a transparent `snafu(display("{}"), source)` in upload.rs?